### PR TITLE
Puppet 6 issue and knockout_prefix problem

### LINKED ISSUE #111: --knock-out-prefix "--" knocks out valid content of yum::gpgkeys
Most of the yum gpg keys start with something like '-----BEGIN PGP PUBLIC KEY BLOCK-----'
which gets knocked out by --knock-out-prefix "--" when hiera deep merges yum::gpgkeys set outside of the module.   

This module works fine when I change --knock-out-prefix to a different value.

## Affected Puppet, Ruby, OS and module versions/distributions

- Puppet: 4.10
- Ruby:
- Distribution:  EL7
- Module version: 3.1

## How to reproduce (e.g Puppet code you use)

add yum::gpgkey with defined content outside of module bundled hiera.

## What are you seeing
Error: Could not retrieve catalog from remote server: Error 500 on SERVER: Server Error: Evaluation Error: Error while evaluating a Resource Statement, Evaluation Error: Error while evaluating a Function Call, Missing params: $content or $source must be specified at /etc/puppetlabs/code/environments/production/modules/yum/manifests/gpgkey.pp:49:18

### LINKED ISSUE #120: In Puppet 6, remove_undef_values doesn't work as expected
## Affected Puppet, Ruby, OS and module versions/distributions

- Puppet: 6.0.0
- Ruby: 2.0.0p648
- Distribution: RHEL 7
- Module version: 3.1.0

## How to reproduce (e.g Puppet code you use)

1. Include the module in Puppet 6, any way you'd like.

## What are you seeing

The puppet run fails due to trying to get the "length" member from a Nil object (because remove_undef_values does not remove the undef (Nil) value).

## What behaviour did you expect instead

Puppet runs without errors.

## Output log
Server Error: Evaluation Error: Error while evaluating a Function Call, undefined method 'length' for nil:NilClass (file: /etc/puppetlabs/code/environments/production/modules/yum/manifests/init.pp, line: 222, column: 20) on node mynode.com.

## Any additional information you'd like to impart
I'm submitting a PR to fix this issue by removing undefined values with a filter.

### PR DESCRIPTION
**Pull request by avidspartan1**

### CODE REVIEW

**Comment by jocelynthode** [comment]
I was able to bypass that when prefixing in my hiera file the yum::gpgkeys with "---" just above.

Do we have an ETA on this PR being merged and a new version release @bastelfreak ?

File: manifests/init.pp [hunk 1]
```diff
@@ -216,7 +216,7 @@
       true    => '--keepdevel',
       default => undef,
     },
-  ])
+  ].filter |$val| { $val =~ NotUndef }
```
Comment (bastelfreak) [inline manifests/init.pp]: oh nice, I didn't know that `NotUndef` is a thing!

**Comment by bastelfreak** [comment]
Hi, thanks for the PR.I'm not sure if this a breaking change or not. Could you explain a bit in which setup the old code break? Could you provide this as an acceptance test?

**Comment by avidspartan1** [comment]
The delete_undef_values code broke in an earlier version of Puppet 6--it looks like it was addressed in a later release. However, the knockout_prefix issue is still a problem. I'll try to find some time to update the PR.

**Comment by a-zen** [comment]
I'm facing the same issue by simply using the yum module without any configuration on a clean centos7 system. The patch from @bastelfreak fixes the issue.

**Comment by TJM** [comment]
I am glad you have this fixed already, cause I was just about to dig into this one, and it was making me dizzy already. :)

File: manifests/init.pp (hunk 1 above)
Comment (TJM) [reply manifests/init.pp]: er, maybe I am just mis-reading this, but wouldn't it have been easier to just set the default for $keep_kernel_devel to `''` (empty string)? That is the only potentially undefined value there, right?

**Comment by larsnaesbye** [comment]
Great! :-) Shouldn't this warrant a release? I don't mind patching locally, but it'd be easier to upgrade the module since Puppet 6 still acts up with 3.1.1.

**Comment by dgoldsmith** [comment]
PR #121 may have been merged on 3/16/2019 but no new release tags have been published since then.  Version 3.1.1 is still the latest tagged release and it does not include the gpgkeys knockout prefix fix.

**Comment by bastelfreak** [comment]
I'm doing a new release at https://github.com/voxpupuli/puppet-yum/pull/134

**Comment by dgoldsmith** [comment]
Thanks, that fixed it.

David Goldsmith

**Pull request merged**